diff --git a/admin/settings/grades.php b/admin/settings/grades.php
index 90bc811..7503c34 100644
--- a/admin/settings/grades.php
+++ b/admin/settings/grades.php
@@ -23,6 +23,10 @@ if (has_capability('moodle/grade:manage', $systemcontext)
     $temp = new admin_settingpage('gradessettings', new lang_string('generalsettings', 'grades'), 'moodle/grade:manage');
     if ($ADMIN->fulltree) {
 
+        // BEGIN LSU Course Category Editable Patch
+        $temp->add(new admin_setting_configcheckbox('grade_coursecateditable', new lang_string('coursecateditable', 'grades'), new lang_string('coursecateditable_help', 'grades'), 1));
+        // END LSU Course Category Editable Patch
+
         // new CFG variable for gradebook (what roles to display)
         $temp->add(new admin_setting_special_gradebookroles());
 
diff --git a/grade/edit/tree/category_form.php b/grade/edit/tree/category_form.php
index 98a18a4..97b0844 100644
--- a/grade/edit/tree/category_form.php
+++ b/grade/edit/tree/category_form.php
@@ -391,7 +391,17 @@ class edit_category_form extends moodleform {
 
             // If it is a course category and its fullname is ?, show an empty field
             if ($grade_category->is_course_category() && $mform->getElementValue('fullname') == '?') {
-                $mform->setDefault('fullname', '');
+                // BEGIN LSU Course Category Editable Patch
+                $editable = (bool) get_config('moodle', 'grade_coursecateditable');
+
+                // If it is a course category and its fullname is ?, show an editable empty field
+                if ($editable && $mform->getElementValue('fullname') == '?') {
+                    $mform->setDefault('fullname', '');
+                } else if (!$editable) {
+                    $mform->setDefault('fullname', $COURSE->fullname);
+                    $mform->hardFreeze('fullname');
+                }
+                // END LSU Course Category Editable Patch
             }
             // remove unwanted aggregation options
             if ($mform->elementExists('aggregation')) {
diff --git a/lang/en/grades.php b/lang/en/grades.php
index 584bf4c..d40676a 100644
--- a/lang/en/grades.php
+++ b/lang/en/grades.php
@@ -813,6 +813,8 @@ $string['writinggradebookinfo'] = 'Writing gradebook settings';
 $string['xml'] = 'XML';
 $string['yes'] = 'Yes';
 $string['yourgrade'] = 'Your grade';
+$string['coursecateditable'] = 'Course category editable';
+$string['coursecateditable_help'] = 'Disabling will make the course category name an uneditable field.';
 
 // Deprecated since 3.0.
 $string['categoriesanditems'] = 'Categories and items';
