diff --git a/admin/settings/grades.php b/admin/settings/grades.php
index 6d12d99..ee81bac 100644
--- a/admin/settings/grades.php
+++ b/admin/settings/grades.php
@@ -68,6 +68,8 @@ if (has_capability('moodle/grade:manage', $systemcontext)
 
         $temp->add(new admin_setting_special_gradelimiting());
 
+        $temp->add(new admin_setting_configcheckbox('grade_report_nocalculations', get_string('nocalculations', 'grades'), get_string('nocalculations_help', 'grades'), 0));
+
         $temp->add(new admin_setting_special_gradepointmax());
 
         $temp->add(new admin_setting_special_gradepointdefault());
diff --git a/grade/edit/tree/lib.php b/grade/edit/tree/lib.php
index c05eb92..638bfe8 100644
--- a/grade/edit/tree/lib.php
+++ b/grade/edit/tree/lib.php
@@ -55,6 +55,9 @@ class grade_edit_tree {
     public function __construct($gtree, $moving=false, $gpr) {
         global $USER, $OUTPUT, $COURSE;
 
+        $system_default = get_config('moodle', 'grade_report_nocalculations');
+        $this->disable_calculations = get_user_preferences('grade_report_nocalculations', $system_default);
+
         $this->gtree = $gtree;
         $this->moving = $moving;
         $this->gpr = $gpr;
@@ -138,7 +141,9 @@ class grade_edit_tree {
             $actions .= $this->gtree->get_edit_icon($element, $this->gpr);
         }
 
-        $actions .= $this->gtree->get_calculation_icon($element, $this->gpr);
+        if (!$this->disable_calculations) {
+            $actions .= $this->gtree->get_calculation_icon($element, $this->gpr);
+        }
 
         if ($element['type'] == 'item' or ($element['type'] == 'category' and $element['depth'] > 1)) {
             if ($this->element_deletable($element)) {
diff --git a/grade/report/grader/lib.php b/grade/report/grader/lib.php
index c91394a..3eac62f 100644
--- a/grade/report/grader/lib.php
+++ b/grade/report/grader/lib.php
@@ -1480,10 +1480,13 @@ class grade_report_grader extends grade_report {
         $lockunlockicon      = '';
 
         if (has_capability('moodle/grade:manage', $this->context)) {
-            if ($this->get_pref('showcalculations')) {
-                $editcalculationicon = $this->gtree->get_calculation_icon($element, $this->gpr);
+            if (!get_config('moodle', 'grade_report_nocalculations')){
+                if ($this->get_pref('showcalculations')) {
+                    $editcalculationicon = $this->gtree->get_calculation_icon($element, $this->gpr);
+                }
             }
 
+
             if ($this->get_pref('showeyecons')) {
                 $showhideicon = $this->gtree->get_hiding_icon($element, $this->gpr);
             }
diff --git a/grade/report/grader/preferences_form.php b/grade/report/grader/preferences_form.php
index ba2cb24..5d10f5e 100644
--- a/grade/report/grader/preferences_form.php
+++ b/grade/report/grader/preferences_form.php
@@ -55,7 +55,10 @@ class grader_report_preferences_form extends moodleform {
         if (has_capability('moodle/grade:manage', $context)) {
 
             $preferences['prefshow'] = array();
-            $preferences['prefshow']['showcalculations']  = $checkbox_default;
+            // if no calculations setting is turned on, don't show this option
+            if(!get_config('moodle', 'grade_report_nocalculations')){
+                $preferences['prefshow']['showcalculations']  = $checkbox_default;
+            }
             $preferences['prefshow']['showeyecons']       = $checkbox_default;
             if ($canviewhidden) {
                 $preferences['prefshow']['showaverages']  = $checkbox_default;
diff --git a/grade/report/grader/settings.php b/grade/report/grader/settings.php
index d1b02af..0ebbc1f 100644
--- a/grade/report/grader/settings.php
+++ b/grade/report/grader/settings.php
@@ -55,8 +55,12 @@ if ($ADMIN->fulltree) {
     $settings->add(new admin_setting_configcheckbox('grade_report_enableajax', get_string('enableajax', 'grades'),
                                                 get_string('enableajax_help', 'grades'), 0));
 
-    $settings->add(new admin_setting_configcheckbox('grade_report_showcalculations', get_string('showcalculations', 'grades'),
+    if(!get_config('moodle', 'grade_report_nocalculations')){
+        $settings->add(new admin_setting_configcheckbox('grade_report_showcalculations', get_string('showcalculations', 'grades'),
                                                 get_string('showcalculations_help', 'grades'), 0));
+    } else {
+        $settings->add(new admin_setting_configempty('grade_report_showcalculations', get_string('showcalculations', 'grades'), get_string('showcalculationsWITH_NO_CALC_help', 'grades'), 0));
+    }
 
     $settings->add(new admin_setting_configcheckbox('grade_report_showeyecons', get_string('showeyecons', 'grades'),
                                                 get_string('showeyecons_help', 'grades'), 0));
diff --git a/lang/en/grades.php b/lang/en/grades.php
index 9ea00eb..fafec1b 100644
--- a/lang/en/grades.php
+++ b/lang/en/grades.php
@@ -596,8 +596,11 @@ $string['showallhidden'] = 'Show hidden';
 $string['showallstudents'] = 'Show all Students';
 $string['showaverages'] = 'Show column averages';
 $string['showaverages_help'] = 'If enabled, the grader report will contain an additional row displaying the average (mean) for each category and grade item.';
+$string['nocalculations'] = 'Disable calculations';
+$string['nocalculations_help'] = 'Do not allow custom calculations anywhere on the site, if an instructor has already created a calculation this will make it impossible for the instructor to change the calculation in the future';
 $string['showcalculations'] = 'Show calculations';
 $string['showcalculations_help'] = 'If enabled, when editing is turned on, a calculator icon is shown for each grade item and category, with tool tips over calculated items and a visual indicator that a column is calculated.';
+$string['showcalculationsWITH_NO_CALC_help'] = '<strong>Calculations have been turned off in the "<a href = "' . $CFG->wwwroot . '/admin/settings.php?section=gradessettings#admin-grade_report_nocalculations">' . $string['generalsettings'] . '</a>" area</strong>. To access this setting please disable the "' . $string['nocalculations'] . '" setting.<br /><br />If "' . $string['showcalculations'] . '" is enabled and editing is turned on within the ' . $string['graderreport'] . ', a calculator icon is shown for each ' . $string['gradeitem'] . ' and ' . $string['category'] . ', with tool tips over calculated items and a visual indicator that a column is calculated.';
 $string['showeyecons'] = 'Show show/hide icons';
 $string['showeyecons_help'] = 'If enabled, when editing is turned on, a show/hide icon is shown for each grade for controlling its visibility to the student.';
 $string['showgroups'] = 'Show groups';
