diff --git a/admin/settings/grades.php b/admin/settings/grades.php
index 6d12d99..dcae382 100644
--- a/admin/settings/grades.php
+++ b/admin/settings/grades.php
@@ -135,6 +135,16 @@ if (has_capability('moodle/grade:manage', $systemcontext)
         $temp->add(new admin_setting_configselect('grade_displaytype', new lang_string('gradedisplaytype', 'grades'),
                                                   new lang_string('gradedisplaytype_help', 'grades'), GRADE_DISPLAY_TYPE_REAL, $display_types));
 
+        $temp->add(new admin_setting_configcheckbox('grade_item_manual_recompute',
+            new lang_string('gradeitemmanualrecompute', 'grades'),
+            new lang_string('gradeitemmanualrecompute_help', 'grades'), 0));
+
+        if ($CFG->grade_item_manual_recompute) {
+            $temp->add(new admin_setting_configcheckbox('manipulate_categories',
+                new lang_string('manipulatecategories', 'grades'),
+                new lang_string('manipulatecategories_help', 'grades'), 0));
+        }
+
         $temp->add(new admin_setting_configselect('grade_decimalpoints', new lang_string('decimalpoints', 'grades'),
                                                   new lang_string('decimalpoints_help', 'grades'), 2,
                                                   array( '0' => '0',
diff --git a/grade/report/grader/db/upgrade.php b/grade/report/grader/db/upgrade.php
new file mode 100644
index 0000000..fdfe40f
--- /dev/null
+++ b/grade/report/grader/db/upgrade.php
@@ -0,0 +1,35 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Upgrade code for gradebook grader report.
+ *
+ * @package   gradereport_grader
+ * @copyright 2013 Moodle Pty Ltd (http://moodle.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+function xmldb_gradereport_grader_upgrade($oldversion) {
+    global $CFG, $DB;
+    // Set rawgrade on manual grade items so that multiplicator and offset
+    // works on them like any other grade item.
+    $sql = "UPDATE  {grade_grades} gr, {grade_items} gi
+            SET     gr.rawgrade = gr.finalgrade
+            WHERE   gi.id = gr.itemid AND gi.itemtype = 'manual' AND gr.rawgrade IS NULL";
+    $DB->execute($sql);
+    upgrade_plugin_savepoint(true, 2014051201, 'gradereport', 'grader');
+    return true;
+}
diff --git a/grade/report/grader/lib.php b/grade/report/grader/lib.php
index c91394a..0c72a34 100644
--- a/grade/report/grader/lib.php
+++ b/grade/report/grader/lib.php
@@ -945,6 +945,12 @@ class grade_report_grader extends grade_report {
 
                 } else if ($USER->gradeediting[$this->courseid]) {
 
+                    // Editing means user edit manual item raw
+                    if ($item->is_manual_item() and $CFG->grade_item_manual_recompute) {
+                        $gradeval = $grade->rawgrade;
+                    }
+
+
                     if ($item->scaleid && !empty($scalesarray[$item->scaleid])) {
                         $scale = $scalesarray[$item->scaleid];
                         $gradeval = (int)$gradeval; // scales use only integers
diff --git a/grade/report/grader/version.php b/grade/report/grader/version.php
index 9b31002..354c9a9 100644
--- a/grade/report/grader/version.php
+++ b/grade/report/grader/version.php
@@ -24,6 +24,6 @@
 
 defined('MOODLE_INTERNAL') || die();
 
-$plugin->version   = 2014051200;        // The current plugin version (Date: YYYYMMDDXX)
+$plugin->version   = 2014051201;        // The current plugin version (Date: YYYYMMDDXX)
 $plugin->requires  = 2014050800;        // Requires this Moodle version
 $plugin->component = 'gradereport_grader'; // Full name of the plugin (used for diagnostics)
diff --git a/lang/en/grades.php b/lang/en/grades.php
index 9ea00eb..68ec46c 100644
--- a/lang/en/grades.php
+++ b/lang/en/grades.php
@@ -689,6 +689,8 @@ $string['verbosescales'] = 'Verbose scales';
 $string['viewbygroup'] = 'Group';
 $string['viewgrades'] = 'View grades';
 $string['warningexcludedsum'] = 'Warning: excluding of grades is not compatible with sum aggregation.';
+$string['manipulatecategories'] = 'Manipulate category totals';
+$string['manipulatecategories_help'] = 'Curve and offset category totals';
 $string['weight'] = 'weight';
 $string['weightcourse'] = 'Use weighted grades for course';
 $string['weightedascending'] = 'Sort by weighted percent ascending';
@@ -700,6 +702,8 @@ $string['weights'] = 'Weights';
 $string['weightsedit'] = 'Edit weights and extra credits';
 $string['weightuc'] = 'Weight';
 $string['writinggradebookinfo'] = 'Writing gradebook settings';
+$string['gradeitemmanualrecompute'] = 'Manual item recompute';
+$string['gradeitemmanualrecompute_help'] = 'Allow the gradebook to automatically adjust the raw grade boundaries when editing the real grade boundaries for Manual items.<br /><strong>When enabled, you will have the option of manipulating category totals.</strong>';
 $string['xml'] = 'XML';
 $string['yes'] = 'Yes';
 $string['yourgrade'] = 'Your grade';
diff --git a/lib/grade/grade_item.php b/lib/grade/grade_item.php
index 3f8b273..05b9f26 100644
--- a/lib/grade/grade_item.php
+++ b/lib/grade/grade_item.php
@@ -702,6 +702,14 @@ class grade_item extends grade_object {
                     continue;
                 }
 
+                // Manual item rawgrade might be recomputed
+                if ($this->is_manual_item() and $CFG->grade_item_manual_recompute) {
+                    $maxscale = ($this->grademax / $grade->rawgrademax);
+                    $grade->rawgrademax = $this->grademax;
+                    $grade->rawgrademin = $this->grademin;
+                    $grade->rawgrade = $this->bounded_grade($grade->rawgrade);
+                }
+
                 $grade->finalgrade = $this->adjust_raw_grade($grade->rawgrade, $grade->rawgrademin, $grade->rawgrademax);
 
                 if (grade_floats_different($grade_record->finalgrade, $grade->finalgrade)) {
@@ -967,7 +975,17 @@ class grade_item extends grade_object {
      * @return bool
      */
     public function is_raw_used() {
-        return ($this->is_external_item() and !$this->is_calculated() and !$this->is_outcome_item());
+        global $CFG;
+        if($CFG->manipulate_categories) {
+            $manipulatable_item = ($this->is_category_item() or $this->is_course_item());
+        } else {
+            $manipulatable_item = NULL;
+        }
+        if($CFG->grade_item_manual_recompute) {
+            return ($this->is_manual_item() or $this->is_external_item() or $manipulatable_item and !$this->is_calculated() and !$this->is_outcome_item());
+        } else {
+            return ($this->is_external_item() and !$this->is_calculated() and !$this->is_outcome_item());
+        }
     }
 
     /**
@@ -1537,6 +1555,14 @@ class grade_item extends grade_object {
             return false;
         }
 
+        // Manual Item raw-grade support
+        if ($this->is_manual_item()) {
+            return $this->update_raw_grade(
+                $userid, $finalgrade, $source, $feedback, $feedbackformat,
+                $usermodified, null, null, $grade
+            );
+        }
+
         $oldgrade = new stdClass();
         $oldgrade->finalgrade     = $grade->finalgrade;
         $oldgrade->overridden     = $grade->overridden;
