diff --git a/admin/settings/grades.php b/admin/settings/grades.php
index 6d12d99..cdaacc1 100644
--- a/admin/settings/grades.php
+++ b/admin/settings/grades.php
@@ -104,6 +104,9 @@ if (has_capability('moodle/grade:manage', $systemcontext)
 
         $options = array(0 => new lang_string('no'), 1 => new lang_string('yes'));
 
+        $temp->add(new admin_setting_configcheckbox('grade_overridecat', new lang_string('overridecat', 'grades'),
+            new lang_string('overridecat_help', 'grades'), 1));
+
         $defaults = array('value'=>1, 'forced'=>false, 'adv'=>true);
         $temp->add(new admin_setting_gradecat_combo('grade_aggregateonlygraded', new lang_string('aggregateonlygraded', 'grades'),
                     new lang_string('aggregateonlygraded_help', 'grades'), $defaults, $options));
diff --git a/grade/report/grader/lib.php b/grade/report/grader/lib.php
index c91394a..c126222 100644
--- a/grade/report/grader/lib.php
+++ b/grade/report/grader/lib.php
@@ -147,6 +147,9 @@ class grade_report_grader extends grade_report {
         $this->setup_groups();
         $this->setup_users();
         $this->setup_sortitemid();
+
+        $this->overridecat = (bool)get_config('moodle', 'grade_overridecat');
+
     }
 
     /**
@@ -1471,7 +1474,17 @@ class grade_report_grader extends grade_report {
         // Init all icons
         $editicon = '';
 
-        if ($element['type'] != 'categoryitem' && $element['type'] != 'courseitem') {
+        $editable = true;
+
+        if ($element['type'] == 'grade') {
+            $item = $element['object']->grade_item;
+
+            if ($item->is_course_item() or $item->is_category_item()) {
+                $editable = $this->overridecat;
+            }
+        }
+
+        if ($element['type'] != 'categoryitem' && $element['type'] != 'courseitem' &&$editable) {
             $editicon = $this->gtree->get_edit_icon($element, $this->gpr);
         }
 
diff --git a/lib/grade/grade_grade.php b/lib/grade/grade_grade.php
index 8763fdd..19cfd17 100644
--- a/lib/grade/grade_grade.php
+++ b/lib/grade/grade_grade.php
@@ -243,6 +243,10 @@ class grade_grade extends grade_object {
             return false;
         }
 
+        if ($grade_item->is_course_item() or $grade_item->is_category_item()) {
+            return (bool)get_config('moodle', 'grade_overridecat');
+        }
+
         return true;
     }
 
diff --git a/lib/grade/grade_item.php b/lib/grade/grade_item.php
index 3f8b273..2679146 100644
--- a/lib/grade/grade_item.php
+++ b/lib/grade/grade_item.php
@@ -949,7 +949,13 @@ class grade_item extends grade_object {
      * @return bool
      */
     public function is_overridable_item() {
-        return !$this->is_outcome_item() and ($this->is_external_item() or $this->is_calculated() or $this->is_course_item() or $this->is_category_item());
+        if ($this->is_course_item() or $this->is_category_item()) {
+            $overridable = (bool) get_config('moodle', 'grade_overridecat');
+        } else {
+            $overridable = false;
+        }
+
+        return !$this->is_outcome_item() and ($this->is_external_item() or $this->is_calculated() or $overridable);
     }
 
     /**
